File: stripe/src/test/java/com/stripe/android/TelemetryClientUtilTest.java
package com.stripe.android;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.support.annotation.NonNull;

import androidx.test.core.app.ApplicationProvider;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import java.util.Map;
import java.util.Objects;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class TelemetryClientUtilTest {

    @Mock
    private PackageManager mPackageManager;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
    }

    @Test
    public void initWithAppContext_shouldSucceed() {
        assertFalse(new TelemetryClientUtil(ApplicationProvider.getApplicationContext())
                .createTelemetryMap()
                .isEmpty());
    }

    @Test
    public void createTelemetryMap_returnsHasExpectedEntries() {
        final Context context = ApplicationProvider.getApplicationContext();
        final Map<String, Object> telemetryMap = new TelemetryClientUtil(
                new FakeUidSupplier(),
                context.getResources().getDisplayMetrics(),
                "package_name",
                context.getPackageManager(),
                "-5"
        )
                .createTelemetryMap();
        assertEquals(5, telemetryMap.size());

        final Map firstMap = Objects.requireNonNull((Map) telemetryMap.get("a"));
        assertEquals(4, firstMap.size());

        assertEquals("Android 9 REL 28", getSingleValue(firstMap, "d"));
        assertEquals("-5", getSingleValue(firstMap, "g"));
        assertEquals(8, Objects.requireNonNull((Map) telemetryMap.get("b")).size());
    }

    @Test
    public void createTelemetryMap_withVersionName_includesVersionName()
            throws PackageManager.NameNotFoundException {
        final PackageInfo packageInfo = new PackageInfo();
        packageInfo.versionName = "version_name";

        when(mPackageManager.getPackageInfo("package_name", 0))
                .thenReturn(packageInfo);

        final Context context = ApplicationProvider.getApplicationContext();
        final Map<String, Object> telemetryMap = new TelemetryClientUtil(
                new FakeUidSupplier(),
                context.getResources().getDisplayMetrics(),
                "package_name",
                mPackageManager,
                "-5"
        )
                .createTelemetryMap();

        final Map secondMap = Objects.requireNonNull((Map) telemetryMap.get("b"));
        assertEquals(9, secondMap.size());
        assertEquals("version_name", secondMap.get("l"));
    }

    @NonNull
    private Object getSingleValue(@NonNull Map map, @NonNull String key) {
        return ((Map) map.get(key)).get("v");
    }
}

File: stripe/src/test/java/com/stripe/android/TelemetryClientUtilTest.kt
package com.stripe.android

import android.content.Context
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import androidx.test.core.app.ApplicationProvider
import com.google.common.truth.Truth.assertThat
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import java.util.UUID
import kotlin.test.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class TelemetryClientUtilTest {

    private val packageManager: PackageManager = mock()
    private val context: Context = ApplicationProvider.getApplicationContext()
    private val clientFingerprintDataStore = FakeClientFingerprintDataStore(
        muid = MUID,
        sid = SID
    )

    @Test
    fun initWithAppContext_shouldSucceed() {
        assertThat(TelemetryClientUtil(context).createTelemetryMap())
            .isNotEmpty()
    }

    @Test
    fun createTelemetryMap_returnsHasExpectedEntries() {
        val telemetryMap = TelemetryClientUtil(
            FakeUidSupplier(),
            context.resources.displayMetrics,
            "package_name",
            context.packageManager,
            "-5",
            clientFingerprintDataStore
        )
            .createTelemetryMap()
        assertThat(telemetryMap)
            .hasSize(5)

        val firstMap = telemetryMap["a"] as Map<*, *>
        assertThat(firstMap)
            .hasSize(4)

        assertThat(getSingleValue(firstMap, "d"))
            .isEqualTo("Android 9 REL 28")
        assertThat(getSingleValue(firstMap, "g"))
            .isEqualTo("-5")

        val secondMap = telemetryMap["b"] as Map<*, *>
        assertThat(secondMap)
            .hasSize(9)
        assertThat(secondMap["d"])
            .isEqualTo(MUID.toString())
        assertThat(secondMap["e"])
            .isEqualTo(SID.toString())
    }

    @Test
    fun createTelemetryMap_withVersionName_includesVersionName() {
        val packageInfo = PackageInfo().also {
            it.versionName = "version_name"
        }

        whenever(packageManager.getPackageInfo("package_name", 0))
            .thenReturn(packageInfo)

        val telemetryMap = TelemetryClientUtil(
            FakeUidSupplier(),
            context.resources.displayMetrics,
            "package_name",
            packageManager,
            "-5",
            clientFingerprintDataStore
        )
            .createTelemetryMap()

        val secondMap = telemetryMap["b"] as Map<*, *>
        assertThat(secondMap)
            .hasSize(10)
        assertThat(secondMap["l"])
            .isEqualTo("version_name")
    }

    private fun getSingleValue(map: Map<*, *>, key: String): Any? {
        return (map[key] as Map<*, *>)["v"]
    }

    private companion object {
        private val MUID = UUID.randomUUID()
        private val SID = UUID.randomUUID()
    }
}
